File: agent/src/life_update_agent/worker.py
"""Layer 4 orchestrator - runs only when the machine is idle and under low
load. Pulls unprocessed raw_events, sessionizes them, summarizes each
session via Presidio + phi3:mini, writes the result to the sync queue, then
hands off to Layer 5 to push anything queued.
"""

from __future__ import annotations

import base64
import json
import logging
import threading
import uuid
from datetime import datetime, timezone

from life_update_agent import db
from life_update_agent.capture.frame_queue import FrameQueue
from life_update_agent.capture.idle_monitor import is_safe_to_run_inference
from life_update_agent.config import Settings
from life_update_agent.inference.cluster import sessionize
from life_update_agent.inference.ollama_client import OllamaError, describe_image
from life_update_agent.inference.summarize import summarize_session
from life_update_agent.redaction.scanner import scan
from life_update_agent.sync.queue import sync_pending

logger = logging.getLogger(__name__)

CHECK_INTERVAL_SECONDS = 60.0
DEVICE_ID_NAMESPACE = uuid.UUID("6f7f9e2c-6c1a-4c2b-9a1e-3d6b6c2f9a11")

VISION_DESCRIBE_PROMPT = (
    "Describe what is being worked on in this screenshot in one or two plain "
    "sentences. Focus on the task or problem, not the UI chrome."
)


def _fetch_unprocessed_events() -> list[dict]:
    with db.get_conn() as conn:
        rows = conn.execute(
            "SELECT * FROM raw_events WHERE processed_at IS NULL ORDER BY ts ASC"
        ).fetchall()
    return [dict(r) for r in rows]


def _mark_processed(event_ids: list[int]) -> None:
    if not event_ids:
        return
    with db.get_conn() as conn:
        conn.executemany(
            "UPDATE raw_events SET processed_at = ? WHERE id = ?",
            [(datetime.now(timezone.utc).isoformat(), eid) for eid in event_ids],
        )


def _enqueue_portfolio_event(event: dict) -> None:
    event_id = str(uuid.uuid5(DEVICE_ID_NAMESPACE, f"{event['started_at']}:{event['ended_at']}"))
    with db.get_conn() as conn:
        conn.execute(
            """INSERT INTO portfolio_event_queue
               (id, started_at, ended_at, project, category, focus_score, apps_used_json, summary, created_at)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
               ON CONFLICT(id) DO NOTHING""",
            (
                event_id,
                event["started_at"],
                event["ended_at"],
                event["project"],
                event["category"],
                event["focus_score"],
                json.dumps(event["apps_used"]),
                event["summary"],
                datetime.now(timezone.utc).isoformat(),
            ),
        )


def _record_screen_text(ts: str, app_name: str | None, title: str | None, text: str) -> None:
    with db.get_conn() as conn:
        conn.execute(
            "INSERT INTO raw_events (ts, kind, app_name, window_title, extra_json) VALUES (?, ?, ?, ?, ?)",
            (ts, "screen_text", scan(app_name), scan(title), scan(text)),
        )


def process_pending_frames(settings: Settings, frame_queue: FrameQueue) -> int:
    """Drains screenshots queued by screen_watcher.py for the vision-model
    path (too slow to OCR inline - see screen_watcher.py). Each frame's
    image is discarded immediately after the description comes back,
    whether that succeeds or fails."""
    frames = frame_queue.drain()
    described = 0

    for frame in frames:
        try:
            image_b64 = base64.b64encode(frame.png_bytes).decode()
            description = describe_image(settings.ollama_host, settings.vision_engine, image_b64, VISION_DESCRIBE_PROMPT)
        except OllamaError:
            logger.warning("vision description failed for a queued frame, discarding", exc_info=True)
            continue

        if description:
            _record_screen_text(frame.ts, frame.app_name, frame.title, description)
            described += 1

    return described


def run_once(settings: Settings) -> int:
    """Runs a single inference pass. Returns the number of sessions processed.

    The worker only runs once the machine has been idle past the threshold,
    so every event fetched here belongs to a *completed* period of activity
    - nothing is still "in progress". That means sessions `sessionize()`
    drops as too short/noisy can be marked processed immediately (they're
    genuinely low-signal, not incomplete); only a live Ollama failure should
    leave events unprocessed for a retry next cycle.
    """
    events = _fetch_unprocessed_events()
    if not events:
        return 0

    sessions = sessionize(events)
    attempted_ids = {e["id"] for session in sessions for e in session}
    noise_ids = [e["id"] for e in events if e["id"] not in attempted_ids]
    _mark_processed(noise_ids)

    processed_count = 0

    for session in sessions:
        result = summarize_session(
            session, settings.ollama_host, settings.ollama_model, settings.presidio_spacy_model
        )
        if result is None:
            logger.warning("skipping session (ollama unavailable or bad output), will retry next cycle")
            continue

        _enqueue_portfolio_event(result)
        _mark_processed([e["id"] for e in session])
        processed_count += 1

    if processed_count:
        sync_pending(settings)

    return processed_count


def run(settings: Settings, stop_event: threading.Event, frame_queue: FrameQueue | None = None) -> None:
    while not stop_event.is_set():
        if is_safe_to_run_inference(settings.idle_threshold_minutes, settings.cpu_load_ceiling_percent):
            if frame_queue is not None and len(frame_queue):
                try:
                    process_pending_frames(settings, frame_queue)
                except Exception:
                    logger.error("processing pending vision frames failed", exc_info=True)

            try:
                count = run_once(settings)
                if count:
                    logger.info("inference pass processed %d session(s)", count)
            except Exception:
                logger.error("inference pass failed", exc_info=True)

        stop_event.wait(CHECK_INTERVAL_SECONDS)

File: app-shell/src-tauri/src/agent/worker.rs
//! Layer 4 orchestrator - runs only when the machine is idle and under low
//! load. Sessionizes unprocessed raw events, summarizes each session via
//! the selected engine, queues the results, and syncs anything unsent.

use super::frame_queue::FrameQueue;
use super::summarize::{summarize_session, RelatedSession, SummaryEngine};
use super::{apple_ai, cluster, db, idle, redaction::scan, sync, AgentConfig};
use base64::Engine as _;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

const CHECK_INTERVAL_SECONDS: u64 = 60;

const VISION_DESCRIBE_PROMPT: &str = "Describe what is being worked on in this screenshot in one or two plain sentences. Focus on the task or problem, not the UI chrome.";

fn describe_image_ollama(host: &str, model: &str, png: &[u8]) -> Result<String, String> {
    let b64 = base64::engine::general_purpose::STANDARD.encode(png);
    let client = reqwest::blocking::Client::builder()
        .timeout(Duration::from_secs(120))
        .build()
        .map_err(|e| e.to_string())?;
    let resp = client
        .post(format!("{}/api/generate", host.trim_end_matches('/')))
        .json(&serde_json::json!({
            "model": model,
            "prompt": VISION_DESCRIBE_PROMPT,
            "images": [b64],
            "stream": false,
            "keep_alive": 0,
        }))
        .send()
        .map_err(|e| format!("ollama unreachable: {e}"))?;
    let body: serde_json::Value = resp.json().map_err(|e| e.to_string())?;
    Ok(body.get("response").and_then(|v| v.as_str()).unwrap_or("").trim().to_string())
}

pub fn process_pending_frames(cfg: &AgentConfig, frames: &FrameQueue) {
    for frame in frames.drain() {
        match describe_image_ollama(&cfg.ollama_host, &cfg.vision_engine, &frame.png_bytes) {
            Ok(description) if !description.is_empty() => {
                db::insert_raw(
                    &cfg.db_path,
                    &frame.ts, // original capture time, not processing time
                    "screen_text",
                    scan(frame.app_name.as_deref()).as_deref(),
                    scan(frame.title.as_deref()).as_deref(),
                    None,
                    scan(Some(&description)).as_deref(),
                );
            }
            Ok(_) => {}
            Err(e) => log::warn!("vision description failed for a queued frame, discarding: {e}"),
        }
    }
}

/// The gate that keeps guesswork off the timeline: a project named after
/// one of the apps used means the model had nothing real to go on.
pub fn project_is_tool_named(project: &str, apps: &[String]) -> bool {
    let p = project.trim().to_lowercase();
    p == "untitled session" || apps.iter().any(|a| a.trim().to_lowercase() == p)
}

/// One inference pass; returns sessions processed. The machine has been
/// idle past the threshold, so everything fetched belongs to a *completed*
/// period of activity - sessions dropped as too short are genuinely noise
/// and are marked processed; only an engine failure leaves events pending.
pub fn run_once(cfg: &AgentConfig) -> usize {
    let events = db::fetch_unprocessed(&cfg.db_path);
    if events.is_empty() {
        return 0;
    }

    let sessions = cluster::sessionize(events.clone(), cluster::DEFAULT_GAP_MINUTES, cluster::DEFAULT_MIN_SESSION_SECONDS);
    let attempted: std::collections::HashSet<i64> =
        sessions.iter().flatten().map(|e| e.id).collect();
    let noise: Vec<i64> = events.iter().map(|e| e.id).filter(|id| !attempted.contains(id)).collect();
    db::mark_processed(&cfg.db_path, &noise);

    let engine = if cfg.model == apple_ai::APPLE_ENGINE {
        SummaryEngine::Apple { helper: &cfg.helper_path }
    } else {
        SummaryEngine::Ollama { host: &cfg.ollama_host, model: &cfg.model }
    };

    // Memory: this user's most recent summarized sessions from the LOCAL
    // store - which the two-way sync keeps consistent with the web, so
    // web-side edits and deletions are reflected here too.
    let related: Vec<RelatedSession> = super::recent_sessions(10)
        .unwrap_or_default()
        .into_iter()
        .filter(|s| !s.held) // never feed held guesswork back as memory
        .take(5)
        .map(|s| RelatedSession { project: s.project, summary: s.summary, ended_at: s.ended_at })
        .collect();

    let mut processed = 0;
    for session in &sessions {
        match summarize_session(session, &engine, &related, None) {
            Some(mut draft) => {
                // Quality gate: tool-named project = the model was guessing.
                // One corrective retry; if still guessing, HOLD it locally
                // for review instead of syncing guesswork to the timeline.
                let mut held = false;
                if project_is_tool_named(&draft.project, &draft.apps_used) {
                    let correction = format!(
                        "The previous attempt named the project \"{}\" - that is a tool or placeholder, never a project. Look again at titles, files and commits for the real name; if none is visible, use a short plain description of the activity instead.",
                        draft.project
                    );
                    if let Some(second) = summarize_session(session, &engine, &related, Some(&correction)) {
                        draft = second;
                    }
                    held = project_is_tool_named(&draft.project, &draft.apps_used);
                }
                sync::enqueue(&cfg.db_path, &draft, held);
                let ids: Vec<i64> = session.iter().map(|e| e.id).collect();
                db::mark_processed(&cfg.db_path, &ids);
                processed += 1;
            }
            None => {
                log::warn!("skipping session (engine unavailable or bad output), will retry next cycle");
            }
        }
    }
    processed
}

pub fn run(cfg: Arc<AgentConfig>, frames: Arc<FrameQueue>, stop: Arc<AtomicBool>) {
    let mut ticks = CHECK_INTERVAL_SECONDS; // first check soon after start
    while !stop.load(Ordering::Relaxed) {
        std::thread::sleep(Duration::from_secs(1));
        ticks += 1;
        if ticks < CHECK_INTERVAL_SECONDS {
            continue;
        }
        ticks = 0;

        // Sync runs on the steady timer, NOT behind the idle gate - a
        // handful of small POSTs is cheap, and gating it meant nothing
        // reached life-update.com until the user stepped away for 3+
        // minutes. Only AI inference needs the idle/low-load gate.
        sync::sync_pending(&cfg.db_path, &cfg.api_url, &cfg.token);
        // Two-way: pull web-side edits/deletions back into the local store
        // on the same cadence, so the app (and the model's memory, which
        // reads locally) always reflects what the user curated on the web.
        sync::pull_remote_sessions(&cfg.db_path, &cfg.api_url, &cfg.token);

        if !idle::is_safe_to_run_inference(cfg.idle_threshold_minutes, cfg.cpu_load_ceiling_percent) {
            continue;
        }

        if frames.len() > 0 && crate::vision_models::is_ollama_backed(&cfg.vision_engine) {
            process_pending_frames(&cfg, &frames);
        }

        let count = run_once(&cfg);
        if count > 0 {
            log::info!("inference pass processed {count} session(s)");
            sync::sync_pending(&cfg.db_path, &cfg.api_url, &cfg.token);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::project_is_tool_named;

    #[test]
    fn tool_named_projects_are_caught() {
        let apps = vec!["Warp".to_string(), "Google Chrome".to_string()];
        assert!(project_is_tool_named("Warp", &apps));
        assert!(project_is_tool_named("google chrome", &apps));
        assert!(project_is_tool_named("Untitled session", &apps));
        assert!(!project_is_tool_named("AcornTracker", &apps));
    }
}
